### PR TITLE
Version the graph examples

### PR DESCRIPTION
**Pull request by nirvdrum**
This is a fixed branch for #84. I had merged that PR already before realizing that I lost the Ruby graphs for GraalVM 21.2.0. Since those cannot be regenerated without a patched Graal, we want to retain the originals. I reverted the PR on _main_ and reworked the branch. This PR supplants #84 (see original for more context).

### CODE REVIEW

File: tools/generate-examples.rb [hunk 1]
```diff
@@ -0,0 +1,428 @@
+#!/usr/bin/env ruby
+# frozen_string_literal: true
+
+require "fileutils"
+require "pathname"
+
+VERBOSE = false
+EXAMPLES_DIR = File.expand_path("#{__dir__}/../examples", __dir__)
```
Comment (rwstauner) [inline tools/generate-examples.rb:8]: I'd write this as 
```suggestion
EXAMPLES_DIR = File.expand_path("../examples", __dir__)
```
As is, I think the second arg will be unused

File: tools/generate-examples.rb [hunk 2]
```diff
@@ -0,0 +1,428 @@
...
+    ] + args,
+    { err: [:child, :out] },
+  )
+
+  yield pipe
+
+  sleep(3)
+  Process.kill("KILL", pipe.pid)
+  Process.wait(pipe.pid)
+end
```
Comment (rwstauner) [inline tools/generate-examples.rb:228]: There's quite a bit more functionality duplicated across the methods that you could move to this function.
```suggestion
def run_command(graalvm, args, options = {})
  if VERBOSE
    log(graalvm, "  Command: #{args.join(" ")}")
  end

  FileUtils.rm_rf("./graal_dumps")

  pipe = IO.popen(args, options)

  yield pipe

  sleep(3)
  Process.kill("KILL", pipe.pid)
  Process.wait(pipe.pid)
end

def run_java(graalvm, bin, dump_level, *args)
  run_command(graalvm, [
    "#{bin}/java", \
    "-XX:+PrintCompilation",
    "-Dgraal.PrintGraphWithSchedule=true",
    "-Dgraal.Dump=:#{dump_level}",
  ] + args)
end

def run_js(graalvm, bin, *args)
  run_command(
    graalvm,
    [
      "#{bin}/node",
      "--experimental-options",
      "--engine.CompileOnly=fib",
      "--engine.TraceCompilation",
      "--engine.MultiTier=false",
      graalvm.options_version == 1 ? "--engine.Inlining=false" : "--compiler.Inlining=false",
      "--vm.Dgraal.Dump=Truffle:1",
    ] + args,
    { err: [:child, :out] },
  )
end

def run_ruby(graalvm, bin, *args)
  run_command(
    graalvm,
    [
      "#{bin}/ruby",
      "--jvm",
      "--experimental-options",
      "--engine.TraceCompilation",
      graalvm.options_version == 1 ? "--engine.NodeSourcePositions" : "--compiler.NodeSourcePositions",
      "--engine.MultiTier=false",
      "--vm.Dgraal.Dump=Truffle:1",
    ] + args,
    { err: [:child, :out] },
  )
en
```

**Review by rwstauner — APPROVED** [review]
A couple of notes but looks fine.

File: tools/generate-examples.rb (hunk 2 above)
Comment (nirvdrum) [reply tools/generate-examples.rb:228]: I'm okay with the duplication. All I wanted was a way to log the command easily. I'm already a bit uneasy with the multiple levels required to execute a command. I think the simplicity of the original bash-like implementation was a lot easier to follow.

**Pull request merged**